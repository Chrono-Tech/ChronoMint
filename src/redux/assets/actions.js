import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const ASSETS_LIST = 'assets/LIST'

export const listAssets = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getAssetsManagerDAO()
  dispatch({type: ASSETS_LIST, list: await dao.getList()})
}
