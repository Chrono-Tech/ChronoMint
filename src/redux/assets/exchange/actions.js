import type ExchangeManagerDAO from 'dao/ExchangeManagerDAO'
import type ExchangeModel from 'models/ExchangeModel'
import type TokenModel from 'models/TokenModel'
import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const EXCHANGE_MANAGER_LIST = 'exchangeManager/LIST'
export const EXCHANGE_MANAGER_SET = 'exchangeManager/SET'
export const EXCHANGE_MANAGER_REMOVE = 'exchangeManager/REMOVE'

const getDAO = async (): Promise<ExchangeManagerDAO> => await contractsManagerDAO.getExchangeManagerDAO()

export const listExchanges = () => async (dispatch) => {
  const dao = await getDAO()
  dispatch({type: EXCHANGE_MANAGER_LIST, list: await dao.getList()})
}

export const addExchange = (address) => async () => {
  const exchange = new ExchangeModel({address, isFetching: true})
  dispatch({type: EXCHANGE_MANAGER_SET, exchange})
  try {
    const dao = await getDAO()
    await dao.addExchange(address)
  } catch (e) {
    // no rollback
  }
}

export const createExchange = (token: TokenModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.createExchange(token)
  } catch (e) {
    // no rollback
  }
}

export const editExchange = (exchange: ExchangeModel, newExchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.editExchange(exchange, newExchange)
  } catch (e) {
    // no rollback
  }
}

export const removeExchange = (exchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.removeExchange(exchange)
  } catch (e) {
    // no rollback
  }
}

// TODO @bshevchenko: MINT-230 AssetsManager
// export const addOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.addOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
//
// export const removeOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.removeOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
