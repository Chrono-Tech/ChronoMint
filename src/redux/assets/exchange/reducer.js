import Immutable from 'immutable'
import * as a from './actions'

const initialState = {
  list: new Immutable.Map(),
  isFetched: false
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case a.EXCHANGE_MANAGER_LIST:
      return {
        ...state,
        list: action.list,
        isFetched: true
      }
    default:
      return state
  }
}

export default reducer
