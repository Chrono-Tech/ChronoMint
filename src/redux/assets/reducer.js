import Immutable from 'immutable'
import * as a from './actions'

const initialState = {
  list: new Immutable.Map(),
  isFetched: false
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case a.ASSETS_LIST:
      return {
        ...state,
        list: action.list,
        isFetched: true
      }
    default:
      return state
  }
}

export default reducer
