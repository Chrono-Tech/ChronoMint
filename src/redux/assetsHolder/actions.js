import { DUCK_SESSION } from '@/redux/session/actions'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import TIMEHolderDAO from 'dao/TIMEHolderDAO'
import Amount from 'models/Amount'
import ApprovalNoticeModel from 'models/notices/ApprovalNoticeModel'
import TokenModel from 'models/tokens/TokenModel'
import { DUCK_TOKENS } from 'redux/tokens/actions'
import tokenService, { EVENT_NEW_TOKEN } from 'services/TokenService'

export const DUCK_ASSETS_HOLDER = 'assetsHolder'

export const TIME_HOLDER_INIT = 'timeHolder/INIT'
export const TIME_HOLDER_TIME_ADDRESS = 'timeHolder/timeAddress'
export const TIME_HOLDER_WALLET_ADDRESS = 'timeHolder/timeHolderWalletAddress'
export const TIME_HOLDER_ADDRESS = 'timeHolder/timeHolderAddress'

let timeHolderDAO: TIMEHolderDAO = null

const subscribeOnTokens = (token: TokenModel) => async (dispatch, getState) => {
  if (token.symbol() !== 'TIME') {
    return
  }

  dispatch({ type: TIME_HOLDER_TIME_ADDRESS, address: token.address() })

  // init
  const [ timeHolderAddress, timeHolderWalletAddress ] = await Promise.all([
    timeHolderDAO.getAddress(),
    timeHolderDAO.getWalletAddress(),
  ])

  dispatch({ type: TIME_HOLDER_ADDRESS, address: timeHolderAddress })
  dispatch({ type: TIME_HOLDER_WALLET_ADDRESS, address: timeHolderWalletAddress })

  // TODO @dkchv: review again
  const contractNames = {}
  contractNames[ timeHolderAddress ] = 'TIME Holder'
  ApprovalNoticeModel.setContractNames(contractNames)

  // update allowance
  const timeDAO = tokenService.getDAO(token.id())
  const { account } = getState().get(DUCK_SESSION)
  const [ timeHolderAllowance, timeHolderWalletAllowance ] = await Promise.all([
    timeDAO.getAccountAllowance(timeHolderAddress, account),
    timeDAO.getAccountAllowance(timeHolderWalletAddress, account),
  ])

  console.log('--assetHolder: ', +timeHolderAllowance, +timeHolderWalletAllowance)

  // token = token
  //   .setAllowance(timeHolderAddress, timeHolderAllowance)
  //   .setAllowance(timeHolderWalletAddress, timeHolderWalletAllowance)
}

export const initAssetsHolder = () => async (dispatch, getState) => {
  if (getState().get(DUCK_ASSETS_HOLDER).isInited()) {
    return
  }
  dispatch({ type: TIME_HOLDER_INIT, inInited: true })

  const { account } = getState().get(DUCK_SESSION)
  timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const [ assets, deposit ] = await Promise.all([
    timeHolderDAO.getSharesContract(),
    timeHolderDAO.getDeposit(account),
  ])
  console.log('--deposit', assets, +deposit)

  // subscribe to tokens
  const callback = (token) => dispatch(subscribeOnTokens(token))
  tokenService.on(EVENT_NEW_TOKEN, callback)
  // fetch for existing tokens
  const tokens = getState().get(DUCK_TOKENS)
  tokens.list().forEach(callback)
}

export const depositAsset = (amount: Amount, token: TokenModel) => async (dispatch) => {
  // const amountBN = new BigNumber(amount)

  // dispatch(balanceMinus(amountBN, token))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.deposit(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('deposit error', e.message)
  } finally {
    // compensation for update in watchTransfer
    // dispatch(balancePlus(amount, token))
  }
}

export const withdrawAsset = (amount: Amount, token: TokenModel) => async () => {
  // const amountBN = new BigNumber(amount)

  // dispatch(depositMinus(amountBN))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.withdraw(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('withdraw error', e.message)
  } finally {
    // dispatch(depositPlus(amountBN))
  }
}
