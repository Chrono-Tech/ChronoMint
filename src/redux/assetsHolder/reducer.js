import TimeHolderModel from 'models/timeHolder/TimeHolderModel'
import * as a from './actions'

const initialState = new TimeHolderModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case a.TIME_HOLDER_INIT:
      return state.isInited(true)
    case a.TIME_HOLDER_TIME_ADDRESS:
      return state.timeAddress(action.address)
    case a.TIME_HOLDER_ADDRESS:
      return state.timeHolderAddress(action.address)
    case a.TIME_HOLDER_WALLET_ADDRESS:
      return state.timeHolderWalletAddress(action.address)
    default:
      return state
  }
}
