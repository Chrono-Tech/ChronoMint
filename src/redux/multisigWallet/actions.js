import multisigWalletService from 'services/MultisigWalletService'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import type MultisigWalletDAO from 'dao/MultisigWalletDAO'
import type MultisigWalletModel from 'models/Wallet/MultisigWalletModel'
import type MultisigWalletPendingTxModel from 'models/Wallet/MultisigWalletPendingTxModel'
import TokenModel from 'models/TokenModel'
import type WalletNoticeModel from 'models/notices/WalletNoticeModel'
import { DUCK_SESSION } from 'redux/session/actions'
import { notify } from 'redux/notifier/actions'

export const DUCK_MULTISIG_WALLET = 'multisigWallet'

export const MULTISIG_FETCHING = 'multisig/FETCHING'
export const MULTISIG_FETCHED = 'multisig/FETCHED'

export const MULTISIG_UPDATE = 'multisigWallet/UPDATE'
export const MULTISIG_SELECT = 'multisigWallet/SELECT'
export const MULTISIG_REMOVE = 'multisigWallet/REMOVE'

const updateWallet = (wallet: MultisigWalletModel) => (dispatch) => {
  let updatedWallet = wallet
  if (!wallet.isNew() && !!wallet.transactionHash()) {
    // address arrived, delete temporary hash
    dispatch({ type: MULTISIG_REMOVE, id: wallet.transactionHash() })
    updatedWallet = wallet.transactionHash(null)
  }
  dispatch({ type: MULTISIG_UPDATE, wallet: updatedWallet.isPending(false) })
}

const watchMultisigWallet = (wallet: MultisigWalletModel) => async () => {
  try {
    await multisigWalletService.subscribeToWalletDAO(wallet)
  } catch (e) {
    // eslint-disable-next-line
    console.error('watch error', e.message)
  }
}

export const watchWalletManager = () => async (dispatch, getState) => {
  const dao = await contractsManagerDAO.getWalletsManagerDAO()
  await dao.watchWalletCreate((wallet: MultisigWalletModel, notice: WalletNoticeModel) => {
    dispatch(updateWallet(wallet.isPending(false)))
    dispatch(notify(notice))
    watchMultisigWallet(wallet)
    const wallets = getState().get(DUCK_MULTISIG_WALLET)
    if (wallets.size() === 1) {
      dispatch(selectMultisigWallet(wallets.first()))
    }
  })

  // TODO @dkchv: !!!
  // multisig wallet events
  multisigWalletService.on('OwnerRemoved', (walletId, result) => {
    // eslint-disable-next-line
    console.log('--actions#', result)
  })

  multisigWalletService.on('MultiTransact', (walletId, multisigTransactionModel) => {
    let wallet = getState().get(DUCK_MULTISIG_WALLET).item(walletId)
    const pendingTxList = wallet.pendingTxList().remove(multisigTransactionModel)
    const tokens = wallet.tokens()
    let token:TokenModel = tokens.get(multisigTransactionModel.symbol())
    if (!token) {
      // eslint-disable-next-line
      console.error('token not found', multisigTransactionModel.symbol())
      return
    }
    token = token.updateBalance(false, multisigTransactionModel.value())

    dispatch(updateWallet(wallet.pendingTxList(pendingTxList).tokens(tokens.set(token.id(), token))))
  })

  multisigWalletService.on('SingleTransact', (walletId, result) => {
    // eslint-disable-next-line
    console.log('--actions#', result)
  })

  multisigWalletService.on('Revoke', (walletId, id) => {
    const wallet: MultisigWalletModel = getState().get(DUCK_MULTISIG_WALLET).item(walletId)
    const pendingTxList = wallet.pendingTxList()
    const pendingTx = pendingTxList.item(id).isConfirmed(false)
    dispatch(updateWallet(wallet.pendingTxList(pendingTxList.list(pendingTxList.list().set(id, pendingTx)))))
  })

  multisigWalletService.on('Confirmation', (walletId, id, owner) => {
    if (owner !== getState().get(DUCK_SESSION).account) {
      return
    }

    const wallet: MultisigWalletModel = getState().get(DUCK_MULTISIG_WALLET).item(walletId)
    const pendingTxList = wallet.pendingTxList()
    let pendingTx = pendingTxList.item(id)
    if (!pendingTx) {
      return
    }
    pendingTx = pendingTx.isConfirmed(true)
    dispatch(updateWallet(wallet.pendingTxList(pendingTxList.list(pendingTxList.list().set(id, pendingTx)))))
  })

  multisigWalletService.on('ConfirmationNeeded', (walletId, pendingTxModel: MultisigWalletPendingTxModel) => {
    const wallet: MultisigWalletModel = getState().get(DUCK_MULTISIG_WALLET).item(walletId)
    const pendingTxList = wallet.pendingTxList()
    dispatch(updateWallet(wallet.pendingTxList(pendingTxList.update(pendingTxModel))))
  })

  multisigWalletService.on('Deposit', (walletId, tokenId, amount) => {
    const wallet: MultisigWalletModel = getState().get(DUCK_MULTISIG_WALLET).item(walletId)
    const token: TokenModel = wallet.tokens().get(tokenId)
    dispatch(updateWallet(wallet.tokens(wallet.tokens().set(token.id(), token.updateBalance(true, amount)))))
    // dispatch(notify(notice))
  })
}

export const selectMultisigWallet = (wallet: MultisigWalletModel) => (dispatch) => {
  dispatch({ type: MULTISIG_SELECT, wallet })
}

export const getWallets = () => async (dispatch) => {
  dispatch({ type: MULTISIG_FETCHING })
  const dao = await contractsManagerDAO.getWalletsManagerDAO()
  const wallets = await dao.getWallets()
  const walletsArray = wallets.toArray()

  // watch for every wallet
  for (let wallet of walletsArray) {
    dispatch(watchMultisigWallet(wallet))
  }

  dispatch({ type: MULTISIG_FETCHED, wallets })
  if (wallets.first()) {
    dispatch(selectMultisigWallet(wallets.first()))
  }
}

export const createWallet = (wallet: MultisigWalletModel) => async (dispatch) => {
  try {
    const dao = await contractsManagerDAO.getWalletsManagerDAO()
    const txHash = await dao.createWallet(wallet)
    dispatch(updateWallet(wallet.isPending(true).transactionHash(txHash)))
  } catch (e) {
    // eslint-disable-next-line
    console.error('create wallet error', e.message)
  }
}

export const removeWallet = (wallet: MultisigWalletModel) => async (dispatch, getState) => {
  try {
    const { account } = getState().get(DUCK_SESSION)
    const dao: MultisigWalletDAO = wallet.dao()
    await dao.removeWallet(wallet, account)
    dispatch({ type: MULTISIG_REMOVE, id: wallet.address() })
  } catch (e) {
    // eslint-disable-next-line
    console.error('delete error', e.message)
  }
}

export const addOwner = (wallet: MultisigWalletModel, ownerAddress: string) => async dispatch => {
  dispatch(updateWallet(wallet.isPending(true)))
  try {
    const dao: MultisigWalletDAO = wallet.dao()
    await dao.addOwner(wallet, ownerAddress)
  } catch (e) {
    // eslint-disable-next-line
    console.error('error', e.message)
  }
}

export const removeOwner = (wallet, ownerAddress) => async dispatch => {
  dispatch(updateWallet(wallet.isPending(true)))
  try {
    const dao: MultisigWalletDAO = wallet.dao()
    await dao.removeOwner(wallet, ownerAddress)
  } catch (e) {
    // eslint-disable-next-line
    console.error('error', e.message)
  }
}

export const multisigTransfer = (wallet, token, amount, recipient) => async (dispatch, getState) => {
  try {
    const dao: MultisigWalletDAO = wallet.dao()
    await dao.transfer(wallet, token, amount, recipient)
  } catch (e) {
    // eslint-disable-next-line
    console.error('error', e.message)
  }
}

export const confirmMultisigTx = (wallet, tx: MultisigWalletPendingTxModel) => async (dispatch) => {
  try {
    const dao: MultisigWalletDAO = wallet.dao()
    await dao.confirmPendingTx(tx)
  } catch (e) {
    // eslint-disable-next-line
    console.error('error', e.message)
  }
}

export const revokeMultisigTx = (wallet: MultisigWalletModel, tx: MultisigWalletPendingTxModel) => async (dispatch) => {
  try {
    const dao: MultisigWalletDAO = wallet.dao()
    await dao.revokePendingTx(tx)
  } catch (e) {
    // eslint-disable-next-line
    console.error('error', e.message)
  }
}
