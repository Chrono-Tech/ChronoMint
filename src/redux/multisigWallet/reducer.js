import MultisigWalletCollection from 'models/wallet/MultisigWalletCollection'

import * as a from './actions'

const initialState = new MultisigWalletCollection()

export default (state = initialState, action) => {
  switch (action.type) {
    case a.MULTISIG_INIT:
      return state.isInited(true)
    case a.MULTISIG_FETCHING:
      return state.leftToFetch(action.count)
    case a.MULTISIG_FETCHED:
      return state.itemFetched(action.wallet)
    case a.MULTISIG_UPDATE:
      return state.list(state.list().set(action.wallet.id(), action.wallet))
    case a.MULTISIG_SELECT:
      return state.selected(action.wallet.address())
    case a.MULTISIG_REMOVE:
      return state.list(state.list().remove(action.id))
    case a.MULTISIG_BALANCE:
      return state.balance(action.walletId, action.symbol, action.balance)
    default:
      return state
  }
}
