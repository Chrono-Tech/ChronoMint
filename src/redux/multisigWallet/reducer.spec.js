/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import Immutable from 'immutable'
import TokenModel from 'models/tokens/TokenModel'
import MultisigWalletCollection from 'models/wallet/MultisigWalletCollection'
import MultisigWalletModel from 'models/wallet/MultisigWalletModel'
import * as a from './actions'
import reducer from './reducer'

const wallet1 = new MultisigWalletModel({
  address: 'a1',
})
const wallet2 = new MultisigWalletModel({
  address: 'a2',
})

describe('Multisig Wallet reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual(new MultisigWalletCollection())
  })

  it('should handle MULTISIG_FETCHING', () => {
    expect(reducer(new MultisigWalletCollection(), {
      type: a.MULTISIG_FETCHING,
    })).toMatchSnapshot()
  })

  it('should handle MULTISIG_FETCHED', () => {
    expect(reducer(new MultisigWalletCollection({
      isFetching: true,
    }), {
      type: a.MULTISIG_FETCHED,
      wallets: new Immutable.Map({
        a1: wallet1,
        a2: wallet2,
      }),
    })).toMatchSnapshot()
  })

  it('should handle MULTISIG_UPDATE (add new wallet)', () => {
    expect(reducer(new MultisigWalletCollection({
      list: new Immutable.Map({
        a1: wallet1,
      }),
    }), {
      type: a.MULTISIG_UPDATE,
      wallet: wallet2,
    }))
  })

  it('should handle MULTISIG_UPDATE (update existing wallet)', () => {
    const updatedWallet = new MultisigWalletModel({
      address: 'a1',
      tokens: new Immutable.Map({
        t1: new TokenModel(),
      }),
    })
    expect(reducer(new MultisigWalletCollection({
      list: new Immutable.Map({
        a1: wallet1,
      }),
    }), {
      type: a.MULTISIG_UPDATE,
      wallet: updatedWallet,
    })).toMatchSnapshot()
  })

  it('should handle MULTISIG_SELECT', () => {
    expect(reducer(new MultisigWalletCollection({
      list: new Immutable.Map({
        a1: wallet1,
      }),
    }), {
      type: a.MULTISIG_SELECT,
      wallet: wallet1,
    })).toMatchSnapshot()
  })

  it('should handle MULTISIG_REMOVE', () => {
    expect(reducer(new MultisigWalletCollection({
      list: new Immutable.Map({
        a1: wallet1,
        a2: wallet2,
      }),
    }), {
      type: a.MULTISIG_REMOVE,
      wallet: wallet1,
    })).toMatchSnapshot()
  })
})
