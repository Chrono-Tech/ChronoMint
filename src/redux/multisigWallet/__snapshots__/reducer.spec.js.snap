// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multisig Wallet reducer should handle MULTISIG_FETCHED 1`] = `
{
  "id": null,
  "isFailed": false,
  "isFetched": true,
  "isFetching": false,
  "isPending": false,
  "list": {
    "a1": {
      "address": "a1",
      "dao": null,
      "id": null,
      "is2FA": false,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isMultisig": true,
      "isPending": false,
      "name": "No name",
      "owners": [],
      "pendingTxList": {
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      },
      "requiredSignatures": null,
      "timestamp": null,
      "tokens": {},
      "transactionHash": null,
      "transactions": {
        "endOfList": false,
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      }
    },
    "a2": {
      "address": "a2",
      "dao": null,
      "id": null,
      "is2FA": false,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isMultisig": true,
      "isPending": false,
      "name": "No name",
      "owners": [],
      "pendingTxList": {
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      },
      "requiredSignatures": null,
      "timestamp": null,
      "tokens": {},
      "transactionHash": null,
      "transactions": {
        "endOfList": false,
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      }
    }
  },
  "selected": null,
  "timestamp": null,
  "transactionHash": null
}
`;

exports[`Multisig Wallet reducer should handle MULTISIG_FETCHING 1`] = `
{
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": true,
  "isPending": false,
  "list": {},
  "selected": null,
  "timestamp": null,
  "transactionHash": null
}
`;

exports[`Multisig Wallet reducer should handle MULTISIG_REMOVE 1`] = `
{
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isPending": false,
  "list": {
    "a1": {
      "address": "a1",
      "dao": null,
      "id": null,
      "is2FA": false,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isMultisig": true,
      "isPending": false,
      "name": "No name",
      "owners": [],
      "pendingTxList": {
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      },
      "requiredSignatures": null,
      "timestamp": null,
      "tokens": {},
      "transactionHash": null,
      "transactions": {
        "endOfList": false,
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      }
    },
    "a2": {
      "address": "a2",
      "dao": null,
      "id": null,
      "is2FA": false,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isMultisig": true,
      "isPending": false,
      "name": "No name",
      "owners": [],
      "pendingTxList": {
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      },
      "requiredSignatures": null,
      "timestamp": null,
      "tokens": {},
      "transactionHash": null,
      "transactions": {
        "endOfList": false,
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      }
    }
  },
  "selected": null,
  "timestamp": null,
  "transactionHash": null
}
`;

exports[`Multisig Wallet reducer should handle MULTISIG_SELECT 1`] = `
{
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isPending": false,
  "list": {
    "a1": {
      "address": "a1",
      "dao": null,
      "id": null,
      "is2FA": false,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isMultisig": true,
      "isPending": false,
      "name": "No name",
      "owners": [],
      "pendingTxList": {
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      },
      "requiredSignatures": null,
      "timestamp": null,
      "tokens": {},
      "transactionHash": null,
      "transactions": {
        "endOfList": false,
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      }
    }
  },
  "selected": "a1",
  "timestamp": null,
  "transactionHash": null
}
`;

exports[`Multisig Wallet reducer should handle MULTISIG_UPDATE (update existing wallet) 1`] = `
{
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isPending": false,
  "list": {
    "a1": {
      "address": "a1",
      "dao": null,
      "id": null,
      "is2FA": false,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isMultisig": true,
      "isPending": false,
      "name": "No name",
      "owners": [],
      "pendingTxList": {
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      },
      "requiredSignatures": null,
      "timestamp": null,
      "tokens": {
        "t1": {
          "address": null,
          "allowance": {},
          "balance": "0",
          "dao": null,
          "decimals": null,
          "fee": null,
          "feeAddress": null,
          "icon": null,
          "id": null,
          "isFailed": false,
          "isFetched": false,
          "isFetching": false,
          "isOptional": true,
          "isPending": false,
          "isReissuable": false,
          "managersList": null,
          "name": null,
          "platform": null,
          "symbol": null,
          "timestamp": null,
          "totalSupply": "0",
          "transactionHash": null,
          "url": null,
          "withFee": false
        }
      },
      "transactionHash": null,
      "transactions": {
        "endOfList": false,
        "id": null,
        "isFailed": false,
        "isFetched": false,
        "isFetching": false,
        "isPending": false,
        "list": {},
        "selected": null,
        "timestamp": null,
        "transactionHash": null
      }
    }
  },
  "selected": null,
  "timestamp": null,
  "transactionHash": null
}
`;
